src/Components/Web.JS/src/Boot.Web.ts

```diff
@@ -9,56 +9,121 @@
+function onDescriptorAdded(descriptor: ServerComponentDescriptor | WebAssemblyComponentDescriptor) {
+  switch (descriptor.type) {
+    case 'server':
+      // Start the circuit now so that it's more likely to be ready by the time interactivity actually starts.
+      startCircuitIfNotStarted();
+      addServerDescriptor(descriptor);
+      break;
+    case 'webassembly':
+      startWebAssemblyIfNotStarted();
+      // TODO: WebAssembly.
+      break;
+  }
+}
 
-  if (serverComponents.length) {
-    // TEMPORARY until https://github.com/dotnet/aspnetcore/issues/48763 is implemented
-    // As soon we we see you have interactive components, we'll stop doing enhanced nav even if you don't have an interactive router
-    // This is because, otherwise, we would need a way to add new interactive root components to an existing circuit and that's #48763
-    detachProgressivelyEnhancedNavigationListener();
+function beforeEnhancedNavigation() {
+  isPerformingEnhnacedNavigation = true;
+}
 
-    await startCircuit(webStartOptions?.circuit, serverComponents);
+function beforeDocumentUpdated() {
+  lastActiveElement = document.activeElement;
+}
+
+function afterDocumentUpdated() {
+  if ((lastActiveElement instanceof HTMLElement) && lastActiveElement !== document.activeElement) {
+    lastActiveElement.focus();
   }
 
-  if (webAssemblyComponents.length) {
-    // TEMPORARY until https://github.com/dotnet/aspnetcore/issues/48763 is implemented
-    // As soon we we see you have interactive components, we'll stop doing enhanced nav even if you don't have an interactive router
-    // This is because, otherwise, we would need a way to add new interactive root components to an existing WebAssembly runtime and that's #48763
-    detachProgressivelyEnhancedNavigationListener();
+  lastActiveElement = null;
+  handleUpdatedDescriptors();
+}
+
+function afterEnhancedNavigation() {
+  isPerformingEnhnacedNavigation = false;
+  handleUpdatedDescriptors();
+}
 
-    await startWebAssembly(webStartOptions?.webAssembly, webAssemblyComponents);
+function handleUpdatedDescriptors() {
+  processRemovedServerDescriptors();
+
+  if (!isPerformingEnhnacedNavigation) {
+    // We avoid adding new root components during enhanced navigations.
+
+    // TODO: Handle 'activate' and 'update' seprately.
+    // We might want to update parameters of existing components
+    // without activating new ones.
+    processUpdatedServerDescriptors();
+  }
+}
+
+let circuitStarted = false;
+async function startCircuitIfNotStarted() {
+  if (circuitStarted) {
+    return;
   }
+
+  circuitStarted = true;
+  await startCircuit(webStartOptions?.circuit);
+  handleUpdatedDescriptors();
+}
+
+let webAssemblyStarted = false;
+async function startWebAssemblyIfNotStarted() {
+  if (webAssemblyStarted) {
+    return;
+  }
+
+  webAssemblyStarted = true;
+  await startWebAssembly(webStartOptions?.webAssembly);
+  handleUpdatedDescriptors();
```

Comment:
I know there's a lot of "Server this" and "WebAssembly that" in this file, but that's hopefully going to go away after I get the WebAssembly functionality working and know what the right abstractions are.